src/charm.py

```diff
@@ -0,0 +1,323 @@
+        self._run_generate_synapse()
+
+    @property
+    def _external_hostname(self):
+        """Return the external hostname passed to ingress via the relation."""
+        # It is recommended to default to `self.app.name` so that the external
+        # hostname will correspond to the deployed application name in the
+        # model, but allow it to be set to something specific via config.
+        return self.config["server_name"] or self.app.name
+
+    def _container(self):
+        """Get the Synapse workload container.
+
+        Returns:
+            The pebble instance of the Synapse container.
+        """
+        return self.unit.get_container(self._CONTAINER_NAME)
+
+    def _populate_synapse_env_settings(self):
+        """Populate env settings to pass around."""
+        if os.getenv("SYNAPSE_SERVER_NAME") is None:
+            os.environ["SYNAPSE_SERVER_NAME"] = self.config["server_name"]
+        if os.getenv("SYNAPSE_REPORT_STATS") is None:
+            os.environ["SYNAPSE_REPORT_STATS"] = self.config["report_stats"]
+        pod_config = {
+                "SYNAPSE_SERVER_NAME": os.getenv("SYNAPSE_SERVER_NAME"),
+                "SYNAPSE_REPORT_STATS": os.getenv("SYNAPSE_REPORT_STATS"),
+                "SYNAPSE_NO_TLS": "True",
+                "POSTGRES_DB": self._stored.db_name,
+                "POSTGRES_HOST": self._stored.db_host,
+                "POSTGRES_PORT": self._stored.db_port,
+                "POSTGRES_USER": self._stored.db_user,
+                "POSTGRES_PASSWORD": self._stored.db_password,
+                }
+        return pod_config
+
+    def _register_user(self, event):
+        """Registers a user for usage with Synapse."""
+        Result = collections.namedtuple(
+                "CommandExecResult", "return_code stdout stderr")
+
+        user = event.params["username"]
+        password = event.params["password"]
+        admin = event.params["admin"]
+        admin_switch = "--admin" if admin == "yes" else "--no-admin"
+
+        cmd = ["register_new_matrix_user",
+               "-u", user, "-p", password,
+               admin_switch, "-c", self._SYNAPSE_CONFIG_PATH,
+               "http://localhost:8008"]
+        process = self._container().exec(
+            cmd,
+            working_dir="/data",
+            environment=self._populate_synapse_env_settings(),
+        )
+        try:
+            stdout, stderr = process.wait_output()
+            result = Result(
+                    return_code=0, stdout=stdout, stderr=stderr)
+        except ops.pebble.ExecError as error:
+            result = Result(error.exit_code, error.stdout, error.stderr)
+        return_code = result.return_code
+        logger.debug(
+                  "Run command: %s, return code %s\nstdout: %s\nstderr:%s",
+                  cmd,
+                  return_code,
+                  result.stdout,
+                  result.stderr,
+         )
+
+    def _run_generate_synapse(self):
```

Comment:
When you generate the config with the env vars set, will change the config file to use something like [this](https://github.com/matrix-org/synapse/blob/develop/docker/conf/homeserver.yaml#L53) instead of sqlite3?

Reply:
Yes, exactly, I added a clarifying paragraph in the docstring of the function, thanks!

Reply:
When I checked the Postgresql database, there were no tables and in the pod there was a homeserver.db file . Am I missing something? 🤔

Here is my Juju status:
```
 juju status --relations
Model      Controller       Cloud/Region        Version  SLA          Timestamp
synapsium  github-pr-a25ab  microk8s/localhost  2.9.42   unsupported  10:44:07-03:00

App      Version                       Status  Scale  Charm                     Channel        Rev  Address         Exposed  Message
db       res:postgresql-image@8a72e11  active      1  postgresql-k8s            latest/stable   20  10.152.183.119  no       Pod configured
ingress  25.3.0                        active      1  nginx-ingress-integrator  latest/stable   59  10.152.183.187  no       
synapse                                active      1  synapse                                    3  10.152.183.34   no       

Unit        Workload  Agent  Address      Ports     Message
db/0*       active    idle   10.1.36.234  5432/TCP  Pod configured
ingress/0*  active    idle   10.1.36.228            
synapse/0*  active    idle   10.1.36.241            

Relation provider  Requirer    Interface  Type     Message
db:db              synapse:db  pgsql      regular  
db:peer            db:peer     peer       peer     
```